File: Assets/_Scripts/Views/PuckView.cs
using RageCure.StateUtils;
using Styx.Entities.PuckModule;
using Styx.Managers;
using Styx.Models;
using Styx.States;
using UnityEngine;

namespace Styx.Views
{
    public class PuckView : MonoBehaviour
    {
        [SerializeField]
        private RigidbodyModel _rigidbodyModel;
        [SerializeField]
        private TransformModel _transformModel;

        [SerializeField]
        private PuckLooseState _puckLooseState;
        [SerializeField]
        private PuckCarriedState _puckCarriedState;

        public PuckState PuckState { get; private set; }

        public void Awake()
        {
            _rigidbodyModel = _rigidbodyModel ?? GetComponent<RigidbodyModel>();
            _transformModel = _transformModel ?? GetComponent<TransformModel>();

            _puckLooseState = _puckLooseState ?? GetComponent<PuckLooseState>();
            _puckCarriedState = _puckCarriedState ?? GetComponent<PuckCarriedState>();
        }

        public void Start()
        {
            _rigidbodyModel.Rigidbody = GetComponent<Rigidbody>();
            _transformModel.Transform = transform;

            PuckState = new PuckState
            {
                Puck = new Puck
                {
                    RigidbodyModel = _rigidbodyModel,
                    TransformModel = _transformModel
                },
                States = new Entities.PuckModule.States
                {
                    PuckCarried = _puckCarriedState,
                    PuckLoose = _puckLooseState
                },
                StateMachine = new StateMachine<PuckState>(_puckLooseState)
            };
            PuckManager.Instance.Register(PuckState);
        }

        public void Update()
        {
            PuckState.StateMachine.Update(PuckState);
        }

        public void FixedUpdate()
        {
            _rigidbodyModel.Update();
            PuckState.StateMachine.FixedUpdate(PuckState);
            Debug.Log("Puck state: " + PuckState.StateMachine.CurrentState);
        }
    }
}

File: Assembly-FSharp/Assembly-FSharp/Views/PuckView.fs
namespace Styx.Views

open UnityEngine
open Styx.States.Puck
open Styx.Models
open RageCure.StateUtils
open Styx.Entities.PuckModule
open Styx.Managers

type PuckView () =
    inherit MonoBehaviour ()

    [<SerializeField>]
    let mutable movementModel: MovementModel = Unchecked.defaultof<MovementModel>
    [<SerializeField>]
    let mutable rigidbodyModel: RigidbodyModel = Unchecked.defaultof<RigidbodyModel>
    [<SerializeField>]
    let mutable transformModel: TransformModel = Unchecked.defaultof<TransformModel>

    // **************************************** STATES ***************************************
    [<SerializeField>]
    let mutable puckLooseState: PuckLooseState = Unchecked.defaultof<PuckLooseState>
    [<SerializeField>]
    let mutable puckCarriedState: PuckCarriedState = Unchecked.defaultof<PuckCarriedState>
    // ***************************************************************************************

    let mutable puckState: PuckState = Unchecked.defaultof<PuckState>

    member this.Start() =
        rigidbodyModel.Rigidbody <- this.GetComponent<Rigidbody>()
        transformModel.Transform <- this.GetComponent<Transform>()
        puckState <-
            { puck =
                { rigidbodyModel = rigidbodyModel
                ; transformModel = transformModel
                }
            ; states = 
                { puckLoose     = puckLooseState
                ; puckCarried   = puckCarriedState
                }
            ; stateMachine = new StateMachine<Puck>(puckLooseState)
            }

        puckState
        |> PuckManager.Instance.Register

    member this.Update() =
        puckState.stateMachine.Update(puckState.puck)
        Debug.Log puckState.stateMachine.CurrentState

    member this.FixedUpdate() =
        rigidbodyModel.Update()
        puckState.stateMachine.FixedUpdate(puckState.puck)
